gslib/context_config.py

```diff
@@ -0,0 +1,203 @@
+  match its pre-encapsulation label. It ignores a section without a
+  matching label.
+
+  Args:
+    contents (str): contents of a PEM file.
+    logger (logging.logger): Gsutil logger.
+
+  Returns:
+    A dict of the PEM file sections.
+  """
+  result = {}
+  pem_lines = []
+  pem_section_name = None
+
+  for line in contents.splitlines():
+    line = line.strip()
+    if not line:
+      continue
+
+    begin, end, name = _IsPemSectionMarker(line)
+    if begin:
+      if pem_section_name:
+        logger.warning('Section %s missing end line and will be ignored.' %
+                       pem_section_name)
+      if name in result.keys():
+        logger.warning('Section %s already exists, and the older section will '
+                       'be ignored.' % name)
+      pem_section_name = name
+      pem_lines = []
+    elif end:
+      if not pem_section_name:
+        logger.warning(
+            'Section %s missing a beginning line and will be ignored.' % name)
+      elif pem_section_name != name:
+        logger.warning('Section %s missing a matching end line. Found: %s' %
+                       (pem_section_name, name))
+        pem_section_name = None
+
+    if pem_section_name:
+      pem_lines.append(line)
+      if end:
+        result[name] = '\n'.join(pem_lines) + '\n'
+        pem_section_name = None
+
+  if pem_section_name:
+    logger.warning('Section %s missing an end line.' % pem_section_name)
+
+  return result
+
+
+class _ContextConfig(object):
+  """Represents the configurations associated with context aware access.
+
+  Only one instance of Config can be created for the program.
+  """
+
+  def __init__(self, logger):
+    """Initializes config.
+
+    Args:
+      logger (logging.logger): Gsutil logger.
+    """
+    self.logger = logger
+
+    self.client_cert_path = None
+    self.client_cert_password = None
+    self.use_client_certificate = config.get('Credentials',
+                                             'use_client_certificate',
+                                             None) == 'True'
+
+    atexit.register(self._UnprovisionClientCert)
+
+    if self.use_client_certificate:
+      command_string = config.get('Credentials', 'cert_provider_command', None)
+      if not command_string:
+        raise CertProvisionError('No cert provider detected.')
+
+      self.client_cert_path = os.path.join(gslib.GSUTIL_DIR, 'caa_cert.pem')
+      try:
+        # Certs provisioned using endpoint verification are stored as a
+        # single file holding both the public certificate and the private key.
+        self._ProvisionClientCert(command_string, self.client_cert_path)
+      except CertProvisionError as e:
+        self.logger.error('Failed to provision client certificate: %s' % e)
+
+  def _ProvisionClientCert(self, command_string, cert_path):
+    """Executes certificate provider to obtain client certificate and keys."""
+    # Monkey-patch command line args to get password-protected certificate.
```

Comment:
Monkey-patch?

Reply:
It's a very engineer'y phrase.
In this context, it means "catch the command and add ('patch') the password flag at the end."

I think "monkey" comes from the fact that it's an inelegant way to do things that stems from poor underlying architecture. In this case, it makes sense to monkey-patch, since we're dealing with a "cert_provider_command" provided as a raw string. Note: I have nothing against monkeys and hold them in high esteem.